### PR TITLE
Integrate primitive operations to eta/core

### PR DESCRIPTION
**Pull request by kunyilu**
1. Integrate Dense Optical Flow into ETA
    - Farneback's method
2. Integrate online background subtraction into ETA
    - MOG2 background subtractor
    - KNN background subtractor
3. Integrate edge detection into ETA
    - Canny edge detection
4. Integrate common feature point detection algorithms into ETA
    - Harris corners
    - FAST feature points
    - ORB feature points

### CODE REVIEW

File: LICENSE [hunk 1]
```diff
@@ -1,4 +1,4 @@
-Copyright (c) 2017, Voxel51, LLC
+Copyright (c) 2017, KUNYI
```
Comment (brimoor) [inline LICENSE]: lol! you decided to claim copyrights for the entire ETA library now?

Please revert this

Comment (kunyilu) [reply LICENSE]: Sorry for that. I did not realize it. solved

File: eta/core/primitives.py [hunk 2]
```diff
@@ -0,0 +1,605 @@
...
+        self.poly_n = poly_n
+        self.poly_sigma = poly_sigma
+        self.use_gaussian_filter = use_gaussian_filter
+
+        self._prev_frame = None
+        self._flags = (
+            cv2.OPTFLOW_FARNEBACK_GAUSSIAN if use_gaussian_filter else 0)
+
+    def _process_frame(self, img):
+        curr_frame = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
```
Comment (jasoncorso) [inline eta/core/primitives.py]: ACK.  We recently decided to establish and enforce a contract that all images ETA reads will be converted to RGB from BGR (when ingested as such).  This call violates that contract.  (Of course, this is color-to-gray, so it is not fatal, but nevertheless should be fixed.)

File: eta/core/primitives.py [hunk 3]
```diff
@@ -0,0 +1,605 @@
...
+                # Compute edges
+                edges = self._process_frame(img)
+
+                if masks_path:
+                    # Write edges mask
+                    np.save(masks_path % p.frame_number, edges.astype(np.bool))
+
+                if vid_path:
+                    # Write edges video
+                    p.write(cv2.cvtColor(edges, cv2.COLOR_GRAY2BGR))
```
Comment (jasoncorso) [inline eta/core/primitives.py]: Again color contract issue.

File: eta/core/primitives.py [hunk 4]
```diff
@@ -0,0 +1,605 @@
...
+            k (0.04): Harris detector free parameter
+        '''
+        self.threshold = threshold
+        self.block_size = block_size
+        self.aperture_size = aperture_size
+        self.k = k
+
+    def _process_frame(self, img):
+        # works in OpenCV 3 and OpenCV 2
+        gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
```
Comment (jasoncorso) [inline eta/core/primitives.py]: Again COLOR.

File: eta/core/video.py [hunk 5]
```diff
@@ -27,15 +27,109 @@
 import cv2
 import numpy as np
 
-import eta.core.image as im
 from eta.core import utils
+import eta.core.image as im
```
Comment (jasoncorso) [inline eta/core/video.py]: This does not follow the convention
`from eta.core import utils as etau` or just `import eta.core.utils at etau` and
`import eta.core.image as etai`

File: eta/core/video.py [hunk 6]
```diff
@@ -195,6 +292,12 @@ def __init__(
                 each frame range when the write() method is called. When
                 out_vidpath is None or "", no videos are written
 
+            out_single_vidpath: a path like "/path/to/video.mp4" that specifies
```
Comment (jasoncorso) [inline eta/core/video.py]: I think this should have been added in a different PR; this is rather peculiar functionality and probably deserves more discussion.  Alas, time is not too abundant to do that.

File: eta/core/primitives.py (hunk 2 above)
Comment (brimoor) [reply eta/core/primitives.py]: I will do this in a separate PR. As currently written, ETA loads images in BGR format

File: eta/core/video.py (hunk 5 above)
Comment (brimoor) [reply eta/core/video.py]: This cleanup is coming in the `veh-sim` branch.

File: eta/core/video.py (hunk 6 above)
Comment (brimoor) [reply eta/core/video.py]: It is awkward. But sometimes we want to process sub-clips, in which case we need `"/path/to/video/%05d-%05d.mp4"`, and sometimes we want to process the entire video, in which case we just want to say `"/path/to/video.mp4"`...

The better practice is probably to introduce `VideoClipsProcessor` to handle clips, but alas this works and is backwards compatible for now.

**Pull request merged**